transifex/native/tools/migrations/gettext.py

```diff
@@ -0,0 +1,871 @@
+
+
+class Transformer(object):
+    """Transforms source strings inside Python files from gettext syntax
+    to Transifex Native syntax.
+
+    It also keeps stats about any errors that have occurred.
+
+    It allows clients to register custom module/function paths in order
+    to support gettext module/function wrappers that exist in framework
+    implementations that use native, such as a Django or a Flask SDK.
+    """
+
+    def __init__(self, methods, import_statement):
+        """Constructor.
+
+        :param GettextMethods methods: the gettext methods to transform
+        :param str import_statement: the import statement to use
+            in all migrated files
+        """
+        self.errors = []
+        self._methods = methods
+        self.import_statement = import_statement
+        self._functions = []
+        self.register_functions(*self._methods.all)
+
+    def register_functions(self, *func_details_list):
+        """Register a custom function to be detected during extraction.
+
+        Each arg in `func_paths` must be a tuple, with its first item
+        a string representing the full path of the function
+        (like 'module.deeper_module.func_name') and its second item
+        a list of tuples representing the items.
+
+        Example:
+        (
+            'django.utils.translation.ugettext',
+            [('message', 'string')],
+        )
+        """
+        for func_path, argument_pairs in func_details_list:
+            nodes = func_path.split('.')
+            if len(nodes) < 1:
+                raise ValueError(
+                    'Function path must contain zero or more modules and '
+                    'a function, e.g. "my_module.translate"'
+                )
+
+            self._functions.append(
+                {
+                    'modules': '.'.join(nodes[:-1]),
+                    'function': nodes[-1],
+                    'arguments': argument_pairs,
+                }
+            )
+
+    def transform(self, src, filename=None):
+        """Parse the given Python file string and extract translatable content.
+
+        :param unicode src: a chunk of Python code
+        :param str filename: the filename of the code, i.e. the filename it
+            came from
+        :return: a list of SourceString objects
+        :rtype: list
+        """
+        # Replace utf-8 magic comment, to avoid getting a
+        # "SyntaxError: encoding declaration in Unicode string"
+        src = ENCODING_PATTERN.sub('# ', src)
+        try:
+            tree = ast.parse(src)
+            visitor = CallDetectionVisitor(
+                [
+                    (x['modules'], x['function'])
+                    for x in self._functions
+                ]
+            )
+            visitor.visit(tree)
+            attree = asttokens.ASTTokens(src, tree=tree)
+            file_migration = FileMigration(filename, src)
+
+            last_migrated_char = 0
+
+        except Exception as e:
```

Comment:
This will raise exceptions if it's not valid Python. We are storing the error with `self.errors.append((filename, e))`, so I personally think it does its job fine.

The only change I see is moving these 3 lines after `else`:
```python
attree = asttokens.ASTTokens(src, tree=tree)
file_migration = FileMigration(filename, src)
last_migrated_char = 0
```